tracing-log/src/lib.rs

```diff
@@ -91,6 +93,70 @@ pub trait AsTrace {
     fn as_trace(&self) -> Self::Trace;
 }
 
+pub trait GenericLogEvent<'a> {
+    fn normalized_metadata(&self, fields: &'a LogVisitor) -> Option<Metadata<'a>>;
+    fn is_log(&self) -> bool;
+}
+
+impl<'a> GenericLogEvent<'a> for Event<'a> {
+    fn normalized_metadata(&self, fields: &'a LogVisitor) -> Option<Metadata<'a>> {
+        if self.is_log() {
+            let original = self.metadata();
+
+            Some(Metadata::new(
+                "log event",
+                fields.target.as_ref().map(String::as_str).unwrap_or("log"),
+                original.level().clone(),
+                fields.file.as_ref().map(String::as_str),
+                fields.line.map(|l| l as u32),
+                fields.module_path.as_ref().map(String::as_str),
+                field::FieldSet::new(&[], self.metadata().callsite()),
+                Kind::EVENT,
+            ))
+        } else {
+            None
+        }
+    }
+
+    fn is_log(&self) -> bool {
+        self.metadata().callsite()
+            == match *self.metadata().level() {
+                Level::TRACE => identify_callsite!(TRACE_CS.0),
+                Level::DEBUG => identify_callsite!(DEBUG_CS.0),
+                Level::INFO => identify_callsite!(INFO_CS.0),
+                Level::WARN => identify_callsite!(WARN_CS.0),
+                Level::ERROR => identify_callsite!(ERROR_CS.0),
+            }
+    }
+}
+
+#[derive(Default)]
+pub struct LogVisitor {
+    target: Option<String>,
+    module_path: Option<String>,
+    file: Option<String>,
+    line: Option<u64>,
+}
+
+impl Visit for LogVisitor {
+    fn record_debug(&mut self, _field: &Field, _value: &fmt::Debug) {}
+
+    fn record_u64(&mut self, field: &Field, value: u64) {
+        if field.name() == "log.line" {
+            self.line = Some(value);
+        }
+    }
+
+    fn record_str(&mut self, field: &Field, value: &str) {
+        match field.name() {
+            "log.file" => self.file = Some(value.to_owned()),
+            "log.target" => self.target = Some(value.to_owned()),
+            "log.module_path" => self.module_path = Some(value.to_owned()),
```

Comment:
Also, ideally, it would be good to use `Field`'s `Eq` implementation rather than using string comparisons, as the former should be much more efficient. If the `LogVisitor` was constructed with the event's metadata, we could have it hang onto the `Field`s from the metadata and compare them to the field we're visiting instead...

Reply:
How would we get the field to compare to without doing a string comparison elsewhere? (like `metadata.fields().field()`)

Reply:
I think my earlier comment was a little unclear --- I didn't mean to say that the string comparison could be completely avoided. However, all the field keys for the log callsite for each verbosity level are already stored in a `lazy_static`:
https://github.com/tokio-rs/tracing/blob/115e74b298cae93cf664e91fdd2ece6e464ce17b/tracing-log/src/lib.rs#L170-L187

When constructing the visitor, we could match on the event's level, and use the set of field keys that are stored for the callsite for that verbosity level:
https://github.com/tokio-rs/tracing/blob/115e74b298cae93cf664e91fdd2ece6e464ce17b/tracing-log/src/lib.rs#L190-L201

That way, we could get a very fast integer comparison rather than a (comparatively) slow string comparison.